blazar/plugins/oshosts/host_plugin.py

```diff
@@ -607,6 +607,20 @@ def get_allocations(self, host_id, query):
         allocs = host_allocations.get(host_id, [])
         return {"resource_id": host_id, "reservations": allocs}
 
+    def reallocate_computehost(self, host_id, data):
+        allocations = self.get_allocations(host_id, data)
+
+        for alloc in allocations['reservations']:
+            host_allocation = db_api.host_allocation_get_all_by_values(
+                compute_host_id=host_id,
+                reservation_id=alloc['id'])[0]
+
+            if self._reallocate(host_allocation):
```

Comment:
What happens if this fails? That seems to mean that the host is removed from the aggregate, but is the lease marked as degraded? I ask because after looking at the resource poller it seems like the work to mark nodes as 'not reservable' is done by the poller, not the reallocation mechanism. So by not working with / duplicating logic of the poller, we might be missing some step where we mark the lease as degraded?

Reply:
For example, in the heal_reservations, there is an `else` branch that does this: https://github.com/ChameleonCloud/blazar/blob/a406012696a0b10f294fbf611c0a227d3c0c6d75/blazar/plugins/oshosts/host_plugin.py#L344-L345

Reply:
Other question, what is the second arg to `reservation_update` doing? Is it a set of flags? If so, we should ensure it doesn't override existing flags; rather, it should be effectively a patch. Otherwise by sending in `resources_changed`: True, we might accidentally be removing other flags.

Reply:
Nothing happens if _reallocate fails since it only returns a boolean value. As for flags, I found this documented in resource monitoring and have updated code to reflect it:
```
Heal damaged leases
^^^^^^^^^^^^^^^^^^^

The following boolean flags are introduced for expressing resource conditions.
They are set to False by default:

* Lease

  * degraded

* Reservation

  * missing_resources
  * resources_changed

After updating the status (= *reservable* field) of the failed resource, the
resource plugin discovers leases which suffer from the failure, and tries to
heal the leases. In the healing process, the resource plugin looks for
available resources which satisfy requirements of the lease. Then, it takes
the following actions:

* Case 1. Lease is not started yet

   * Case 1-1. Enough resources are available

     Reserve new resources instead of failure resources. Although resources are
     changed, keep the *resources_changed* flag False because it does not
     affect the lease from the lease owner perspective if it is not started
     yet.

   * Case 1-2.  Alternative resource is not available

     Set the *degraded* flag of the lease and the *missing_resources* flag of
     the reservation to True.

* Case 2. Lease has been already started

   * Case 2-1. Enough resources are available:

     Set the *degraded* flag of the lease and the *resources_changed* flag of
     the reservation to True.

   * Case 2-2.  Alternative resource is not available

     Set the *degraded* flag of the lease and the *missing_resources* flag of
     the reservation to True.

Once the *degraded* and *missing_resources* flags are set True, they are kept
True even if the failed resource is recovered. To make them False, the lease
owner sends an update lease request and requested capacity have to be
satisfied. Note that the *resources_changed* flag never becomes False once it
becomes True. In this case, the *degraded* flag never becomes False, neither.

From the architectural view, resource-dependent tables like the computehosts,
computehost_allocations, computehost_reservations and instance_reservations
tables are updated by the resource plugin. General tables like leases and
reservations tables are updated by the manager.
```